src/java/org/apache/cassandra/index/sai/plan/Operation.java

```diff
@@ -0,0 +1,335 @@
+        {
+            if (node.children().isEmpty())
+                expressions.add(node.expression());
+            else
+            {
+                List<RowFilter.Expression> expressionList = new ArrayList<>();
+                for (Node child : node.children())
+                    doTreeAnalysis(child, expressionList, controller);
+                node.analyze(expressionList, controller);
+            }
+        }
+
+        FilterTree buildFilter(QueryController controller)
+        {
+            analyzeTree(controller);
+            FilterTree tree = filterTree();
+            for (Node child : children())
+                if (child.canFilter())
+                    tree.addChild(child.buildFilter(controller));
+            return tree;
+        }
+    }
+
+    public static abstract class OperatorNode extends Node
+    {
+        List<Node> children = new ArrayList<>();
+
+        @Override
+        public List<Node> children()
+        {
+            return children;
+        }
+
+        @Override
+        public void add(Node child)
+        {
+            children.add(child);
+        }
+    }
+
+    public static class AndNode extends OperatorNode
+    {
+        @Override
+        public void analyze(List<RowFilter.Expression> expressionList, QueryController controller)
+        {
+            expressionMap = analyzeGroup(controller, OperationType.AND, expressionList);
+        }
+
+        @Override
+        FilterTree filterTree()
+        {
+            return new FilterTree(OperationType.AND, expressionMap);
+        }
+
+        @Override
+        RangeIterator rangeIterator(QueryController controller)
+        {
+            RangeIterator.Builder builder = controller.getIndexes(expressionMap.values());
+            for (Node child : children)
+            {
+                boolean canFilter = child.canFilter();
+                if (canFilter)
+                    builder.add(child.rangeIterator(controller));
+            }
+            return builder.build();
+        }
+    }
+
+    public static class ExpressionNode extends Node
+    {
+        RowFilter.Expression expression;
+
+        @Override
+        public void analyze(List<RowFilter.Expression> expressionList, QueryController controller)
+        {
+            expressionMap = analyzeGroup(controller, OperationType.AND, expressionList);
+        }
+
+        @Override
+        FilterTree filterTree()
+        {
+            return new FilterTree(OperationType.AND, expressionMap);
+        }
+
+        public ExpressionNode(RowFilter.Expression expression)
+        {
+            this.expression = expression;
+        }
+
+        @Override
+        public RowFilter.Expression expression()
+        {
+            return expression;
+        }
+
+        @Override
+        RangeIterator rangeIterator(QueryController controller)
+        {
+            assert canFilter();
```

Comment:
nit: Maybe a brief message around what happened?

Reply:
I've put a message on there but I question why we need an assertion here at all. The query will continue quite happily with an empty expression map. How would you feel about getting rid of this assertion?

Reply:
WFM